### PR TITLE
Dependency Injection Trait Locker for Uniques Pallet

### PR DESCRIPTION
**Pull request by HashWarlock**
## Description
Within the RMRK Standard there requires a locking mechanism for the NFT to ensure that when a NFT is `list`ed in the `rmrk-market` pallet, the NFT must be locked and can no longer allow any interactions with the NFT. Since the RMRK pallet extends the `uniques` pallet, implementing the `Locker` trait in the config of `rmrk-core` creates a situation where a user can bypass the `lock` and execute a `do_transfer` directly to the `uniques` pallet. We looked using the `freeze` mechanism for an `asset` instance, but the implementation was not feasible and led to some centralized problems if the RMRK Admin were to be defined as the `freezer` for all NFT collections.

The proposal here is to create a dependency injection trait named `Locker` in the `uniques` pallet config that has a `impl` of `Locker` for `()` that has function named `should_check_lock` that defaults a return of `false`, but can be implemented downstream for a pallet like RMRK to enable a simple locking mechanism to ensure the lock cannot be bypassed by directly calling the `uniques` pallet `do_transfer`. A working example of this in the RMRK pallet is defined [here](https://github.com/rmrk-team/rmrk-substrate/pull/76).

**Note**
Use case defined in RMRK substrate pallet in this PR: https://github.com/rmrk-team/rmrk-substrate/pull/76 where the code currently is pointing to a branch off of the stable branch version of `v0.19.17` with the same changes as this PR. This is my first PR to Substrate so let me know if I may have missed any details and I'll gladly add them. Thank you.

cumulus-companion: https://github.com/paritytech/cumulus/pull/1085

### CODE REVIEW

**Comment by h4x3rotab** [comment]
LGTM

File: frame/uniques/src/impl_locker.rs
Comment (xlc) [inline frame/uniques/src/impl_locker.rs:10]: make it `is_locked`?

File: frame/uniques/src/impl_locker.rs
Comment (xlc) [inline frame/uniques/src/impl_locker.rs:1]: need file header

File: frame/uniques/src/impl_locker.rs
Comment (xlc) [inline frame/uniques/src/impl_locker.rs:19]: why is this needed?

File: frame/uniques/src/impl_locker.rs
Comment (HashWarlock) [reply frame/uniques/src/impl_locker.rs:10]: I can update the naming to `is_locked`.

File: frame/uniques/src/impl_locker.rs
Comment (HashWarlock) [reply frame/uniques/src/impl_locker.rs]: Removed this as it is not needed.

File: frame/uniques/src/impl_locker.rs
Comment (HashWarlock) [reply frame/uniques/src/impl_locker.rs]: Did I update correctly or do I need to create a `locker.rs` for runtime traits? I am not sure.

File: frame/uniques/src/impl_locker.rs
Comment (shawntabrizi) [inline frame/uniques/src/impl_locker.rs:3]:
```suggestion
// Copyright (C) 2022 Parity Technologies (UK) Ltd.
```

File: frame/uniques/src/impl_locker.rs
Comment (shawntabrizi) [inline frame/uniques/src/impl_locker.rs:22]: why is this needed?

File: frame/uniques/src/impl_locker.rs
Comment (shawntabrizi) [inline frame/uniques/src/impl_locker.rs:23]: This trait should probably be placed in `frame_support::traits`

**Review by shawntabrizi — APPROVED** [review]

File: frame/uniques/src/impl_locker.rs
Comment (shawntabrizi) [inline frame/uniques/src/impl_locker.rs:26]: as a warning, the logic in this check must be linear time and consistent for benchmarks to work

File: frame/uniques/src/impl_locker.rs
Comment (HashWarlock) [reply frame/uniques/src/impl_locker.rs:23]: Would it be appropriate for this trait to go into the `misc.rs` file since it is a smaller trait?

Comment (shawntabrizi) [reply frame/uniques/src/impl_locker.rs:23]: probably misc is fine, but please do expose it correctly so i can access from `frame_support::traits`

Comment (HashWarlock) [reply frame/uniques/src/impl_locker.rs:23]: I realized I can do this and looks cleaner:
- Remove file `impl_locker.rs` since `pallet_uniques` doesn't need to implement it
- Add the `impl_locker.rs` code to `misc.rs` in `frame_support::traits`
- Expose `Locker` in the `traits.rs` file in `frame_support` 
- Update the `pallet_uniques` `lib.rs` to include `frame_support::traits::Locker`

File: frame/uniques/src/impl_locker.rs
Comment (HashWarlock) [reply frame/uniques/src/impl_locker.rs]: removed

File: frame/support/src/traits/misc.rs
Comment (kianenigma) [inline frame/support/src/traits/misc.rs:954]:
```suggestion
/// downstream to extend logic of Uniques current functionality.
```

Comment (kianenigma) [reply frame/support/src/traits/misc.rs:954]: If this trait is only useful for the uniques pallet, it should be in `token` folder, not `misc.rs`.

**Review by kianenigma — APPROVED** [review]
Looks broadly good to me, have some minor notes to address before merge.

File: frame/support/src/traits/misc.rs
Comment (HashWarlock) [reply frame/support/src/traits/misc.rs:954]: Would `token/misc.rs` be the best place?

**Comment by HashWarlock** [comment]
Any other updates I need to make to ensure this gets merged?

File: frame/support/src/traits/tokens/misc.rs
Comment (kianenigma) [inline frame/support/src/traits/tokens/misc.rs:195]: these docs should go in the trait, rustdoc in impl is not rendered afaik.

File: frame/support/src/traits/tokens/misc.rs
Comment (kianenigma) [inline frame/support/src/traits/tokens/misc.rs:196]:
```suggestion
pub trait Locker<ClassId, InstanceId> {
	/// Check if the asset should be locked and prevent interactions with the asset from executing.
	fn is_locked(class: ClassId, instance: InstanceId) -> bool;
}

impl<ClassId, InstanceId> Locker<ClassId, InstanceId> for () {
	// Default will be false if not implemented downstream.
	//
	// Note: The logic check in this function must be constant time and consistent for benchmarks
	// to work.
	fn is_locked(_class: ClassId, _instance: InstanceId) -> bool {
```

**Comment by kianenigma** [comment]
Looking at the checks, you need a polkadot companion. Otherwise good to go 👍

**Comment by HashWarlock** [comment]
> Looking at the checks, you need a polkadot companion. Otherwise good to go +1

Would this be the companion you mean? https://github.com/paritytech/cumulus/pull/1085 or I need to add additional one to the polkadot repo as well?

**Comment by kianenigma** [comment]
Yes but you have to put it in the PR description (which I did for you now :) )

**Comment by kianenigma** [comment]
bot merge

**Comment by HashWarlock** [comment]
> Yes but you have to put it in the PR description (which I did for you now :) )

Thank you for the help. Do I need labels and approvals on the cumulus PR as well?

**Comment by kianenigma** [comment]
We have to wait for the new review rules to approve the cumulus companion, sorry for the delay!

**Comment by shawntabrizi** [comment]
bot merge

**Comment by shawntabrizi** [comment]
bot merge

**Comment by shawntabrizi** [comment]
bot rebase

**Comment by gilescope** [comment]
bot merge

**Pull request merged**
